hail/src/main/scala/is/hail/expr/ir/agg/CollectAggregator.scala

```diff
@@ -9,7 +9,8 @@ import is.hail.io.{BufferSpec, InputBuffer, OutputBuffer}
 
 class CollectAggregator(val elemType: PType) extends StagedAggregator {
 
-  val resultType = PArray(elemType, required = false)
+  assert(PType.canonical(elemType) == elemType)
```

Comment:
is this different than checking the virtual types?

Reply:
or are you planning to check requiredeness compatibility through the canonical function?

Reply:
> is this different than checking the virtual types?

Yes! This is very different.

We need this change because the collect aggregator is copying data. The copied data needs to be canonical, and we should fail loudly if this assumption is violated (When we add a type like a SelectedFieldsStruct, for instance).

Reply:
gotcha thanks!

Reply:
In that case, isCanonical?

Reply:
Ah, checked the implementation of isCanonical, it's just doing this basically; we should really have a canonical trait.

Reply:
> we should really have a canonical trait.

This won't work. You could have a canonical type that has a nested non-canonical type.

Reply:
Why wouldn’t it work for that trait to be filled in based on that nesting, lazily?

Reply:
Can you describe what you mean?

Reply:
Sure, if I need to know whether a PType is canonical or not, and that depends on whether it’s children are canonical, I could have a function that evaluates that by reviewing into the children. I could make that run once per ptype by making it lazy. The result would be stored on the ptype,

Reply:
oh, I thought you meant trait as in `PCanonicalArray extends PType with PCanonical` or something, which wouldn't work. You're certainly right that there's a no-allocation strategy for computing this, which I indicated in the comment:

https://github.com/hail-is/hail/blob/1685cba6ec3619fa3aee8f178e6809e78947e005/hail/src/main/scala/is/hail/expr/types/physical/PType.scala#L175